### ISSUE TITLE
Add MapProperty type to support lazy configuration of key/value types

### ISSUE THREAD

**Issue by mkobit** [gradle/gradle#6863]
Provide a similar type to `SetProperty` and `ListProperty` for key-value handling.

### Expected Behavior

Would like the ability to incrementally build a `Map` input using `Provider` types from other lazy configuration. Having a specialized `MapProperty` would make this possible.

* `add(Provider<? extends Map<A, B>)`
* `add(Map<A, B>)`
* possibly `add(A, B)`
* possibly `add(Provider<A>, B)`
* possibly `add(A, Provider<B>)`
* possibly `add(Map.Entry<A, B>)`
* possibly `add(Provider<? extends Map.Entry<A, B>)`

(generics probably off, and API surface area should be considered)

From `kotlin-dsl`, it would be useful to define or generate methods accepting `Pair<A, B>`

### Current Behavior

Need to use something like `Property<Map<A, B>>` or `SetProperty<Map.Entry<A, B>>`.

Using `SetProperty<Map.Entry<A, B>>` it is easiest to piggyback off of the existing [`AbstractMap.SimpleImmutableEntry`](https://docs.oracle.com/javase/8/docs/api/java/util/AbstractMap.SimpleImmutableEntry.html):

```kotlin
import java.util.AbstractMap
val mySetProperty = objects.property()
mySetProperty.add(AbstractMap.SimpleImmutableEntry("first", "second))
val someProperty: Property<String> = objects.property()
mySetProperty.add(someProperty.map { value -> AbstractMap.SimpleImmutableEntry(value, "second) }
```

Using `Property<Map<A, B>>` (untested)

```kotlin
myProperty.set(myProperty.map { it + ("first" to "second") })
```

This approach also allows for removing/mutating elements if needed, but is, IMO, complex.

### Context

Using key/value pairs as input to tasks is fairly commonplace, at least for me. Using the [lazy configuration](https://docs.gradle.org/current/userguide/lazy_configuration.html) is unwieldy.

### Your Environment

Gradle 4.10.2

**Comment by marcphilipp** [comment]
I agree that this would be useful.

@gradle/native Do you already have plans to add that?

**Comment by big-guy** [comment]
@marcphilipp Nothing is planned in the near term, but it's on the radar: https://github.com/gradle/gradle-native/issues/746

**Comment by mkobit** [comment]
Thanks for the link @big-guy - I thought that I remembered an issue for it but couldn't find it before filing.

**Comment by marcphilipp** [comment]
@big-guy Shall we close this one then?

**Comment by mkobit** [comment]
The 2nd example I listed I never tested, but is probably affected by https://github.com/gradle/gradle/issues/7774

**Comment by tkrullmann** [comment]
Hi, I implemented something similar for my [gradle-helm-plugin](https://github.com/unbroken-dome/gradle-helm-plugin), see the code [here](https://github.com/unbroken-dome/gradle-helm-plugin/blob/master/src/main/kotlin/org/unbrokendome/gradle/plugins/helm/util/MapProperty.kt). The code is in Kotlin but I'd be happy to convert this to Java and provide a PR.

I noticed that in 5.x the internals of providers have changed a bit, so now implementing this in a plugin would probably mean having two separate versions for Gradle 4 and Gradle 5 support. That's why I'd really like to see this in the core set of providers too.
